src/tensor_ops/cmp/mod.rs

```diff
@@ -0,0 +1,458 @@
+    ($kernel_op:ty, $try_op:ident, $op:ident, $try_scalar_op:ident, $scalar_op:ident, $doc:expr) => {
+        impl<S: Shape, E: Unit, D: CmpKernel<$kernel_op, E>, T: Tape<D>> Tensor<S, E, D, T> {
+            #[doc = $doc]
+            pub fn $try_op(&self, other: &Self) -> Result<Tensor<S, bool, D, NoneTape>, D::Err> {
+                try_cmp_op(self, other)
+            }
+
+            #[doc = $doc]
+            pub fn $op(&self, other: &Self) -> Tensor<S, bool, D, NoneTape> {
+                self.$try_op(other).unwrap()
+            }
+        }
+
+        impl<S: Shape, E: Unit, D: ScalarCmpKernel<$kernel_op, E>, T: Tape<D>> Tensor<S, E, D, T> {
+            #[doc = $doc]
+            pub fn $try_scalar_op(
+                &self,
+                scalar: E,
+            ) -> Result<Tensor<S, bool, D, NoneTape>, D::Err> {
+                try_scalar_cmp_op(self, scalar)
+            }
+
+            #[doc = $doc]
+            pub fn $scalar_op(&self, scalar: E) -> Tensor<S, bool, D, NoneTape> {
+                self.$try_scalar_op(scalar).unwrap()
+            }
+        }
+    };
+}
+
+impl_cmp_kernel_op!(EqKernelOp, try_eq, eq, try_scalar_eq, scalar_eq, "See [eq]");
+impl_cmp_kernel_op!(NeKernelOp, try_ne, ne, try_scalar_ne, scalar_ne, "See [ne]");
+impl_cmp_kernel_op!(GtKernelOp, try_gt, gt, try_scalar_gt, scalar_gt, "See [gt]");
+impl_cmp_kernel_op!(GeKernelOp, try_ge, ge, try_scalar_ge, scalar_ge, "See [ge]");
+impl_cmp_kernel_op!(LtKernelOp, try_lt, lt, try_scalar_lt, scalar_lt, "See [lt]");
+impl_cmp_kernel_op!(LeKernelOp, try_le, le, try_scalar_le, scalar_le, "See [le]");
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        shapes::{Const, Unit},
+        tensor::*,
+        tests::TestDevice,
+    };
+
+    type TestTensor<const R: usize, const C: usize, E> =
+        Tensor<(Const<R>, Const<C>), E, TestDevice>;
+
+    fn test_cmp<E: Unit, const R: usize, const C: usize, F>(
+        a: [[E; C]; R],
+        b: [[E; C]; R],
+        cmp: F,
+        expected: [[bool; C]; R],
+    ) where
+        F: Fn(&TestTensor<R, C, E>, &TestTensor<R, C, E>) -> [[bool; C]; R],
+    {
+        let dev: TestDevice = Default::default();
+        let a = dev.tensor(a);
+        let b = dev.tensor(b);
+        let r = cmp(&a, &b);
+        assert_eq!(r, expected);
+    }
+
+    fn test_scalar_cmp<E: Unit, const R: usize, const C: usize, F>(
+        a: [[E; C]; R],
+        cmp: F,
+        expected: [[bool; C]; R],
+    ) where
+        F: Fn(&TestTensor<R, C, E>) -> [[bool; C]; R],
+    {
+        let dev: TestDevice = Default::default();
+        let a = dev.tensor(a);
+        assert_eq!(cmp(&a), expected);
+    }
+
+    #[test]
+    fn test_eq() {
+        test_cmp(
+            [[1.0, 2.0, 3.0], [4.0, 5.0, 0.0]],
+            [[0.0, 2.0, -3.0], [4.0, 0.5, -0.0]],
+            |a, b| a.eq(b).array(),
+            [[false, true, false], [true, false, true]],
+        );
+    }
+
+    #[cfg(not(feature = "cuda"))]
```

Comment:
Ahhh this is necessary because cuda kernels are only doing f32 dtype right now? Can you add a `// TODO` comment to remove the feature once they are added?

Reply:
No it's needed because Cuda does not support integers. These tests are making sure that the operations behave as expected for types which are `Unit` but not `Dtype`.